### PR TITLE
[Bugfix] Remove race in fused groupwise RMSNorm quantization

### PR DESCRIPTION
**Pull request by mgoin**
## Purpose

Fix three flaky fused quantized RMSNorm failures observed in the H200 kernel CI job:
https://buildkite.com/vllm/ci/builds/85850#01a04456-3602-4e62-b405-0c986c85cdc3

The two group-64 scale failures were caused by a real shared-memory race in `warpReduceMaxSpecialized`. Each reduction stage read values written by other lanes in the preceding stage without synchronization. CUDA racecheck reported thousands of hazards for the failing specialization.

Replace that reduction with a register-based warp shuffle and write only the final lane-0 result back to shared memory. This supports both CUDA warp-32 and ROCm wave-32/wave-64 paths and also handles groups with more or fewer contributors than the warp width.

The remaining CUDA BF16 per-token FP8 case had bit-identical scales and one isolated output differing by exactly one E4M3 ULP out of 2,103,296 values. Keep scale comparison strict, but permit a bounded number of isolated one-ULP output differences for this specific CUDA/BF16/per-token/FP8 configuration.

## Why this is not a duplicate

I searched open PRs for `fused quant layernorm`, `rms_norm_per_block_quant`, and `warpReduceMaxSpecialized`. No open PR addresses this reduction race or the per-token assertion.

The nearest open work is materially different:

- #45770 adds mixed input/weight dtype support.
- #49639 restores the scalar RMSNorm intermediate rounding boundary.
- #37196 refactors older static quantization kernels.

None changes `quantization/fused_kernels/layernorm_utils.cuh` or removes this shared-memory hazard.

## Performance

Direct H200 measurements timed the full fused op with preallocated outputs. Cold-L2 results used explicit L2 flushing; hot results used ten kernels per CUDA graph replay. Builds were alternated new/old/new or old/new/old to control for drift.

| BF16 + residual case | Cold L2, old -> new | Hot CUDA graph, old -> new |
|---|---:|---:|
| 2048x1024, group-64 INT8 | 22.82 -> 19.39 us (**15.0% faster**) | 17.81 -> 14.36 us (**19.3% faster**) |
| 2048x1024, group-64 FP8 | 24.48 -> 20.96 us (**14.4% faster**) | 19.68 -> 16.16 us (**17.9% faster**) |
| 2048x1024, group-128 FP8 | 22.50 -> 20.03 us (**11.0% faster**) | 17.67 -> 15.16 us (**14.2% faster**) |
| 1x1024, group-64 FP8 | 7.84 -> 7.84 us (no change) | 4.13 -> 4.01 us (**2.9% faster**) |
| 1x5120, group-64 FP8 | 12.45 -> 12.03 us (**3.3% faster**) | 7.77 -> 7.28 us (**6.3% faster**) |
| 1x5120, group-128 FP8 | 11.74 -> 11.52 us (**1.9% faster**) | 7.05 -> 6.77 us (**4.0% faster**) |

CUPTI Python was unavailable locally, so the FlashInfer benchmark helper used its CUDA-event/CUDA-graph fallback.

## Model evaluation

A seeded two-layer `Qwen2ForCausalLM` model-output equivalence check replaced all five RMSNorm outputs with FP8 group quantize/dequantize results and compared the unfused reference against this fused kernel over an 8x32 token batch:

| Group size | Token logits | Top-1 agreement | Max / mean abs logit diff | Mean cosine | Repeat deterministic |
|---|---:|---:|---:|---:|---:|
| 64 | 256 | 100% | 0 / 0 | 1.0 | yes, bitwise |
| 128 | 256 | 100% | 0 / 0 | 1.0 | yes, bitwise |

## Test plan and results

CUDA incremental build:

```bash
cmake --build --preset release --target install
```

Passed.

Original three failing cases:

```bash
.venv/bin/python -m pytest -q \
  'tests/kernels/core/test_fused_quant_layernorm.py::test_rms_norm[False-cuda:0-0-dtype0-True-2048x1024-int8-group-64-tma-0-scale-ub-False]' \
  'tests/kernels/core/test_fused_quant_layernorm.py::test_rms_norm[False-cuda:0-0-dtype0-True-2048x1024-float8_e4m3fn-group-64-tma-0-scale-ub-False]' \
  'tests/kernels/core/test_fused_quant_layernorm.py::test_rms_norm[True-cuda:0-0-dtype0-True-2048x1027-float8_e4m3fn-per-token-tma-0-scale-ub-False]'
```

Result: `3 passed, 14 warnings`.

Additional correctness coverage:

- FP8 group-64 at 1x5120: passed.
- INT8 group-128 at 1x5120: passed.
- 100 repeated 2048x1024 group-64 scale comparisons across INT8 and FP8: all bitwise deterministic.

CUDA racecheck covered both the original group-64 specialization and wider group-64/group-128 configurations:

```text
RACECHECK SUMMARY: 0 hazards displayed (0 errors, 0 warnings)
```

Before this change, the original specialization reported four warning classes and thousands of hazards.

Applicable repository hooks:

```bash
.venv/bin/pre-commit run --files \
  csrc/libtorch_stable/quantization/fused_kernels/layernorm_utils.cuh \
  tests/kernels/core/test_fused_quant_layernorm.py
```

All passed.

## AI assistance

OpenAI Codex assisted with diagnosis, implementation, racecheck analysis, benchmarking, model-output evaluation, test execution, and drafting this description. I reviewed every changed line and verified the commands and results above.

### CODE REVIEW

**Comment by KilaWattCloud** [comment]
Great fix, @mgoin. Eliminating that shared-memory hazard via register warp shuffles (`__shfl_sync`) is a clean way to patch `warpReduceMaxSpecialized` while keeping warp-32/wave-64 portable. Those H200 throughput gains (up to ~19%) on FP8/INT8 fused norm quant are huge for high-concurrency inference workloads.

We’re actually benching low-latency vLLM deployments across multi-region H200/H100 clusters over at **Kilawatt Cloud** to eliminate manual quota friction for automated agent pipelines. 

If you or the team have specific telemetry or orchestration requirements for automated, balance-gated GPU provisioning, we'd love your input on our open RFC thread:
https://github.com/KilaWattCloud-dev/kilawatt-cloud-api/discussions

Awesome work on this kernel patch! 🚀
